### PR TITLE
Add Operator for Apollo

### LINKED ISSUE #4708: [Proposal] apollo-operator design details
This proposal follows https://github.com/apolloconfig/apollo/issues/4696

# Summary
As Cloud Native and Container Orchestration are popular, service lifecycle management has different ways under Kubernetes, it makes service deployment, upgrade and uninstall more easier than before. So for the Apollo Configuration Management System ecosystem, it is the time to change and follow up.

# Motivation
So many users manage Microservices and related systems in Kubernetes, include Apollo, and the traditional system lifecycle management way is complicate in Cloud Native world, so we need to find out a better way to do lifecycle management in Kubernetes.

# Proposal
An Operator extends Kubernetes to automate the management of the entire life cycle of a particular application. Operators serve as a packaging mechanism for distributing applications on Kubernetes, and they monitor, maintain, recover, and upgrade the software they deploy.

### Backend Operator
- For external Mysql DB, you need to create the secret for Apollo-operator using it.

```yaml
cat <<EOF | kubectl apply -f -
apiVersion: v1
kind: Secret
metadata:
  name: myapollodbsecret
type: Opaque
data:
  # KEY: base64-encoded string
  HOST: base64(my-apollometaserver-db)
  PORT: DFlMmU2
  USER_NAME: YWRtaW4=
  PASSWORD: MWYyZDFlMmU2N2Rm
EOF
```
 ### ApolloEnvironment CRD

This CRD response for LCM of Apollo AdminService and Apollo ConfigService, It also has spec to declare which database it uses, and which ApolloPortal CR it uses.
So, when the user apply this CR, the operator will deploy a ConfigService/MetaService and Adminservice and bind them together as a whole.

Besides, operator also provides embedded mysql.
```yaml
cat <<EOF | kubectl apply -f -
apiVersion: k8s.apolloconfig.com/v1
kind: ApolloEnvironment
metadata:
  name: quickstart-apollo-env
spec:
  version: 1.2
  #ingress: need more investigations
  http:
      service:
        spec:
          type: LoadBalancer #  NodePort, ClusterIP(default)
  configServiceCount: 1
  adminServiceCount: 1
  # this embedded mysql only for POC or demo. otherwise, you need to use external db way.
  # suggest no support embed db in this CR, if want to demo use the all in one CR
  mysqlEmbedded: true
  env: one of LOCAL, DEV, BETA, FWS, FAT, UAT, LPT, PRO, TOOLS, UNKNOWN;       
#https://github.com/apolloconfig/apollo/blob/master/apollo-core/src/main/java/com/ctrip/framework/apollo/core/enums/Env.java
  #mysql:
       # the secret of extenal db url and username/password. that means you should create secret before apply AplloEnvironment CRDs. 
       # secretRef :myapollodbsecret
EOF
```

Of course, we also need to detect whether the user is the first time installing or upgrading the version. As Apollo are stateless services, it mainly involves database changes, so, We need to consider the handling of different versions of SQL statements. we can create a job to modify database schemes or tables.

### ApolloPortal CRD
This CRD response for LCM of Apollo Portal and the portal can be used for ApolloEnv. It also has spec to declare which database it uses. And it allows user use one ApolloPortal refs multiple ApolloEnvironment CR instances.

```yaml
cat <<EOF | kubectl apply -f -
apiVersion: k8s.apolloconfig.com/v1
kind: ApolloPortal
metadata:
  name: quickstart
spec:
  version: 0.0.1
  count: 1
  #ingress: need more investigations
  http:
      service:
        spec:
          type: LoadBalancer #  NodePort, ClusterIP(default)
  #mysqlEmbedded: true
  mysql:
    # the secret of extenal db url and username/password. that means you should create secret before apply AplloEnvironment CRDs. 
    secretRef :myapollodbsecret
  apolloEnvRef:
    - name: quickstart-apollo-env
    - name: another-quickstart-apollo-env
EOF
```
### Apollo CRD
This CRD response for LCM a Apollo all in one demo, includes Portal, AdminService, ConfigService and database server, recommending only demonstration usage. So, we can also name it All-In-One(configservice + admin + portal + embedded mysql db) for demo:

```yaml
cat <<EOF | kubectl apply -f -
apiVersion: k8s.apolloconfig.com/v1
kind: Apollo
metadata:
  name: quickstart
spec:
  version: 0.0.1
  count: 1
  #ingress: need more investigations
  http:
      service:
        spec:
          type: LoadBalancer #  NodePort, ClusterIP(default)
EOF
```

## Technology Selection
- Development Language: GO
- Operator dev tool: kubebuilder
- Building tool: Make & Docker
- Installation: Helm3 chart
- Repository: github.com/apolloconfig/apollo-operator
- CI: Github action

more details of this design docs can refs: https://docs.google.com/document/d/1p9OJMpLiZvkeFvTBE8GaFjyPD3RhZf1vxNbr6d_gVqo/edit#

@apolloconfig/pmc Looking forward to your comments, and if there are no objections, I will proceed to the next step of creating the github.com/apolloconfig/apollo-operator repo as our more explicit action items.

### PR DESCRIPTION
**Pull request by CSWYF3634076**
What's the purpose of this PR
Add Operator for Apollo

Which issue(s) this PR fixes:
Fixes https://github.com/apolloconfig/apollo/issues/4708

Proposal: https://dm9sk73nuf.feishu.cn/file/EOV0b6odSocRoZxWAabcicOenlf

Apollo-Operator demo video：https://dm9sk73nuf.feishu.cn/docx/V1oudoUsuo7q0WxPoqLcAMNEnTd

Finished content：portal&environment&allinone create delete update
Unfinished content： 1.default value;  2.using secret instead of partial configmap;  3. log;

### CODE REVIEW

File: api/v1alpha1/apolloportal_types.go [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+	// Ingress is used to specify how ApolloPortal is exposed.
+	// +optional
+	Ingress Ingress `json:"ingress,omitempty"`
+}
+
+type Service struct {
+	Port       int32              `json:"port,omitempty"`
+	TargetPort int32              `json:"targetPort,omitempty"`
+	Type       corev1.ServiceType `json:"type,omitempty"`
+	// TODO 后续看是否必要，不必则删除
```
Comment (JaredTan95) [inline api/v1alpha1/apolloportal_types.go]: all comments should be in English.

File: pkg/reconcile/apollo/deleted.go [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+	"apollo.io/apollo-operator/pkg/utils/naming"
+	"context"
+	"fmt"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	networkingv1 "k8s.io/api/networking/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// DeleteConfigMaps 删除configmap
```
Comment (JaredTan95) [inline pkg/reconcile/apollo/deleted.go]: the same issue, should be in English

File: controllers/apolloenvironment_controller.go [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+// the ApolloEnvironment object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.12.2/pkg/reconcile
+func (r *ApolloEnvironmentReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	// logger := log.FromContext(ctx)
+	log := r.log.WithValues("ApolloPortal", req.NamespacedName)
+	log.Info("进入ApolloPortalReconciler Reconcile")
```
Comment (JaredTan95) [inline controllers/apolloenvironment_controller.go]: all logs should be in English

**Review by JaredTan95 — CHANGES_REQUESTED** [review]

File: go.mod [hunk 4]
```diff
@@ -0,0 +1,83 @@
+module apollo.io/apollo-operator
```
Comment (JaredTan95) [inline go.mod]: how about keeping it same with Apollo config domain `www.apolloconfig.com`, e.g. `module apolloconfig.com/apollo-operator`

**Review by JaredTan95 — CHANGES_REQUESTED** [review]

File: controllers/apolloenvironment_controller.go [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+// SetupWithManager sets up the controller with the Manager.
+func (r *ApolloEnvironmentReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&apolloiov1alpha1.ApolloEnvironment{}).
+		Owns(&corev1.ConfigMap{}).
+		Owns(&corev1.ServiceAccount{}).
+		Owns(&corev1.Service{}).
+		Owns(&appsv1.Deployment{}).
+		Owns(&appsv1.DaemonSet{}).
+		Owns(&appsv1.StatefulSet{}).
```
Comment (JaredTan95) [inline controllers/apolloenvironment_controller.go]: Do we need to watch daemonset and statefulset?

File: controllers/apolloportal_controller.go [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+// SetupWithManager sets up the controller with the Manager.
+func (r *ApolloPortalReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&apolloiov1alpha1.ApolloPortal{}).
+		Owns(&corev1.ConfigMap{}).
+		Owns(&corev1.ServiceAccount{}).
+		Owns(&corev1.Service{}).
+		Owns(&appsv1.Deployment{}).
+		Owns(&appsv1.DaemonSet{}).
+		Owns(&appsv1.StatefulSet{}).
```
Comment (JaredTan95) [inline controllers/apolloportal_controller.go]: the same issue here.

**Review by JaredTan95 — CHANGES_REQUESTED** [review]

**Comment by JaredTan95** [comment]
>  manager E0910 13:42:13.325747       1 reflector.go:138] pkg/mod/k8s.io/client-go@v0.24.2/tools/cache/reflector.go:167: Failed to watch *v1.Ingress: failed to list *v1.Ingress
│ : ingresses.networking.k8s.io is forbidden: User "system:serviceaccount:apollo-operator-system:apollo-operator-controller-manager" cannot list resource "ingresses" in API gro
│ up "networking.k8s.io" at the cluster scope

need to check the permission issue.

**Comment by CSWYF3634076** [comment]
> > manager E0910 13:42:13.325747       1 reflector.go:138] pkg/mod/k8s.io/client-go@v0.24.2/tools/cache/reflector.go:167: Failed to watch *v1.Ingress: failed to list *v1.Ingress
> > │ : ingresses.networking.k8s.io is forbidden: User "system:serviceaccount:apollo-operator-system:apollo-operator-controller-manager" cannot list resource "ingresses" in API gro
> > │ up "networking.k8s.io" at the cluster scope
> 
> need to check the permission issue.

done

**Review by JaredTan95 — APPROVED** [review]
Thank you for your persistence and contribution in contributing such a wonderful component and feature to the apollo community, which makes apollo more possible in the cloud native scenario.

**Comment by nobodyiam** [comment]
Fantastic work! Should we incorporate guidance on configuring the new Apollo CR within the [README.md](https://github.com/apolloconfig/apollo-operator/blob/main/README.md) file?

**Comment by JaredTan95** [comment]
> Fantastic work! Should we incorporate guidance on configuring the new Apollo CR within the [README.md](https://github.com/apolloconfig/apollo-operator/blob/main/README.md) file?

yes,it will be added in following PRs.

**Comment by CSWYF3634076** [comment]
> Fantastic work! Should we incorporate guidance on configuring the new Apollo CR within the [README.md](https://github.com/apolloconfig/apollo-operator/blob/main/README.md) file?

Yes, I will add explanations for each field in yaml later

**Pull request merged**